rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java

```diff
@@ -392,103 +401,299 @@ public final String getName() {
     return this.name;
   }
 
+  public void declareParameter(ParameterVariant parameter) {
+    declareParameter(parameter, new rcl_interfaces.msg.ParameterDescriptor().setName(parameter.getName()));
+  }
+
+  private void internalDeclareParameter(ParameterVariant parameter, rcl_interfaces.msg.ParameterDescriptor descriptor) {
+    if (this.parameters.containsKey(parameter.getName())) {
+      throw new IllegalArgumentException("Parameter is already declared");
+    }
+
+    ParameterAndDescriptor pandd = new ParameterAndDescriptor();
+    pandd.parameter = parameter;
+    pandd.descriptor = descriptor;
+
+    this.parameters.put(parameter.getName(), pandd);
+  }
+
+  public void declareParameter(ParameterVariant parameter, rcl_interfaces.msg.ParameterDescriptor descriptor) {
+    synchronized (parametersMutex) {
+      internalDeclareParameter(parameter, descriptor);
```

Comment:
So I've actually refactored this in 36daa72 to be a lot nicer.  Basically I changed it so that the singular methods always call the plural methods.  Then I take the lock in the plural methods.  That way, we get the atomicity we want, without having the internal method.  The one exception to this is `setParameter/setParameters/setParametersAtomically`, because there we actually do want the locking to be different depending on which variant we are calling.  Take a look at the new code and let me know what you think.

Reply:
LGTM